# Usage
you **MUST** install `snakemake` following the [docs](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html) and install softwares these workflows depend on.

We provide a simple way to install the dependences include snakemake:  

    conda create env -f environment.yml  

If error occuors, check the message, maybe you need to install some dependences by your self.  

# Files
## Files in Snakemake_codes
    Snakemake_codes
    |-- Nanopore_SV.py
    |-- Nanopore_SV.yaml
    |-- Pair-end.py
    |-- Pair-end.yaml
    |-- WGS.py
    |-- WGS.yaml
    |-- single.py
    `-- single.yaml
Files ended with .py is `snakemake` rule for workflow running.  
YAML files contains parameter settings for these workflow, defining locating of FASTQ files, reference, annotations and other setting.

## Files in Data folder
    Data
    |-- Nanopore 
    |   |-- Nano_sub
    |   `-- Nano_sub.fastq.gz
    |-- RNA_seq
    |   |-- RNA_sub_R1.fastq.gz
    |   |-- RNA_sub_R1_fastqc.html
    |   |-- RNA_sub_R2.fastq.gz
    |   `-- RNA_sub_R2_fastqc.html
    `-- WGS
        |-- sub_R1.fastq.gz
        |-- sub_R1_fastqc.html
        |-- sub_R2.fastq.gz
        `-- sub_R2_fastqc.html
1. This Folder contains Sequencing data extract from REAL experiment sequencing reads through `seqtk`, these files are ending by "fastq.gz".
The RNA-seq and WGS data wrere extrackted 15000 reads for every run, and 10000 reads for Nanopore sequencing.  
2. The HTML file were generated by `FastQC`, and Nano_sub folder contains quality report generated by `NanoPlot`.
3. Nano_sub.fastq.gz contains a gene deletion (*ZMO0028*)  in *Zymomonas mobilis* ZM4. RNA_sub_R1 and RNA_sub_R2 is paired RNA-seq data of *Zymomonas mobilis* 3.6M in pH6 . sub_R1 and sub_R2 is paired genome Re-seq data of *Zymomonas mobilis* 3.6M in pH6. ([The work of strain 3.6M have been published](https://biotechnologyforbiofuels.biomedcentral.com/articles/10.1186/s13068-020-01781-1)，you can get RAW sequencing data through PRJNA590883 and PRJNA644990）

Extracted sequence is for TESTING, which contains enough reads to get data but also small enough to save uploading and running time.
# Usage
## Nanopore
For example: 

Fill file locations in Nanopore_SV.yaml

then

    $ cd Snakemake_codes
    $ snakemake -np -c 4 -s Nanopore_SV.py

## WGS
For example:

    $ cd Snakemake_codes
    $ snakemake -np -c 4 -s WGS.py

## RNA-seq
It depends on your sequencing layout to analys RNA-seq data.

Most sequencing data is `Paired-end`, and you can use:

    $ snakemake -np -c 4 -s Pair-end.py

for `Single-end` data, you can use:

    $ snakemake -np -c 4 -s single.py


## Notice:
### threads

`-c 4` set number of threads to use, DO NOT set less than `4`, otherwise you should modify thread setting in snakemake files.

For example, a rule use four threads in Nanopore_SV.py:

    rule Minimap2_map:
    input:
        filted_fq = "Filted/clean.filtlong.fq.gz",
        ref = config["reference"]["Ref"]
    output:
        temp("Map/aln.sam")
    log:
        "log/aln.log"
    threads: 4
    shell:
        "minimap2 --MD -t {threads} -ax map-ont {input.ref} {input.filted_fq} > {output}"

change `threads: 4` to `threads:1`
### GFF and GTF files

1. In GFF files， feature '`Name=***`' is NECESSERY

    for example:

        CP023715.1	feature	gene	147	2516	.	+	.	Name=ZMO1_ZMO0001
        CP023715.1	feature	CDS	147	2516	.	+	0	codon_start=1;Name=ZMO1_ZMO0001;product=TonB-dependent receptor;protein_id=AVZ41384.1;transl_table=11;translation=MKNFIKKGGFLFSFSWFFLQYPVEAATDTVSNTSNNDAIIVTGTRETHKKIRDSLSPIEVLNNRELLETGQTNVTSALAQLVPSITQPAVGQFVAAPTNFVSLRGLNPNQTLVLVNGKRRHNSSFLYIDGFADAATPTDLDLIAPELIDHIEVLKDGAAAQYGSDAIAGVVNIILKSDNHGGSARSQIGQTYAGDGLVGQAGFNKGFKIGHSGFFDVAFDFRHQNHTSRDGIDSRTQRHSLKVVGDPMATRYNLAINAGYDFGNGIEIYTTDSYSHRNSEVNQVYRTADRFPEVYPDGFMPIQKLSENDFSASLGLRGDNALGIHWDVNSTYGGNFIRNDLDKTANLGLYAATGSTPLSVHLNNYSTTQLNNTLDLSKELALPVIYSPLTLAAGFAHRYETYKTGAGDPASYLYGGTQARTGIIPAVAGSHSRNVYSGYLDLSAHLTKRWQVDLAGRYEYYSDFGSTLNGKASTRYDFTDQFALRATFSSGTRAPSLANEYFSTLSVGPDSASGTLGANSAAAKLLGAVPLKPEKATNITAGLVYSPLKNLHFTLDSYQIAIRNRIVSGPGISGEAALAALNAQGVTVASTLEAQNISAWFFTNGASTRTRGLDFTASYHSRFENFGIVDWDIALNINATTIRHVNQLANGMSALNAQTRAYLTSSTPKNRITFGGRWESFSKKWDVSLHEQRFGQTTDEMTWYQGPNAYSMTDFNRIHNHPRWITNLEIGYRPIEKLRVAIGANNLFNAHTTRIPAANGYYGSGKYDGAASQIGVNGGFYYLQTSYQF
        CP023715.1	feature	gene	2538	3653	.	+	.	gene=hisC1;Name=ZMO1_ZMO0002
        CP023715.1	feature	CDS	2538	3653	.	+	0	codon_start=1;gene=hisC1;Name=ZMO1_ZMO0002;product=histidinol-phosphate aminotransferase HisC;protein_id=AVZ41385.1;transl_table=11;translation=MKPLSSINKRQFLSLSALTAFLPMATQVLGKNSGKEASKNLVYANLNENPFPPSPLVKQAIEAELDQIGRYSESAAAQKLVNQIASLENIDPDQIILGEVLGSLGLFLAGIKPDQGNIVYSVPGYTALVDAGKPAGLKGIGIALNKALQNDLPALEKAITAETKAVYLVNPHNPSGTVNHHTDFSEFLARASQKTLIIVDEAYLEYAHFAQTAASFTQKGANVAVFRTLDKIYGLAGMPIGYLLAPKALAASLREAGLGDPHDLGRLTVAAARAALSDQKRRLFIHDYTVKNRVRLTAALDKLKLTHTESESNFVFFKTPIPFSEAREKFYQSGIVVARPFPPLNDWIRLSLGRDEEVSRIITALNRIFGA
2. In GTF files， feature '`gene_id`' and '`transcript_id`' is NECESSERY

    for example:

        CP023715	GeneBank	exon	147	2516	.	+	.	gene_id "ZMO1_ZMO0001";	transcript_id "ZMO1_ZMO0001";
        CP023715	GeneBank	CDS	147	2516	.	+	.	gene_id "ZMO1_ZMO0001";	transcript_id "ZMO1_ZMO0001";


# support

<chenyunhao@stu.hubu.edu.cn>